### PR TITLE
FAI-11519 | Refactor getBoards logic in project-board-filter

### PR DESCRIPTION
**Pull request by matiaslcoulougian**
## Description
Refactor getBoards to request boards by project key, so we avoid using board location which is not populated for Jira Server/Data Center. Also fixed some typos. 

Note: with this change we are increasing the amount of request for getting boards, from one requests (or more depending on the size) for getting all boards to one request (or more) per project to get its boards.

## Type of change
- [x] Bug fix
- [ ] New feature
- [ ] Breaking change

## Related issues

> Fix [#1]() 

## Migration notes

> Describe migration notes if any

## Extra info

> Add any additional information

### CODE REVIEW

File: sources/jira-source/src/project-board-filter.ts [hunk 1]
```diff
@@ -48,45 +48,50 @@ export class ProjectBoardFilter {
...
-              this.farosClient,
-              this.config.graph ?? DEFAULT_GRAPH
-            )
-          : jira.getBoards();
+
+      // Ensure projects is populated
+      await this.getProjects();
+
+      if (this.supportsFarosClient()) {
+        const boards = jira.getBoardsFromGraph(
```
Comment (chalenge) [inline sources/jira-source/src/project-board-filter.ts:56]: `const boards = await jira.getBoardsFromGraph(`

File: sources/jira-source/src/project-board-filter.ts [hunk 2]
```diff
@@ -48,45 +48,50 @@ export class ProjectBoardFilter {
...
+          ) {
+            this.boards.add(board.id.toString());
+          }
         }
       } else {
-        for (const board of this.config.boards) {
-          if (await jira.isBoardInBucket(board)) {
-            await this.maybeIncludeBoard(board);
+        for (const project of this.projects) {
+          for await (const board of await jira.getBoards(project)) {
```
Comment (chalenge) [inline sources/jira-source/src/project-board-filter.ts:79]: remove first await

File: sources/jira-source/src/project-board-filter.ts [hunk 3]
```diff
@@ -48,45 +48,50 @@ export class ProjectBoardFilter {
...
-          ? jira.getBoardsFromGraph(
-              this.farosClient,
-              this.config.graph ?? DEFAULT_GRAPH
-            )
-          : jira.getBoards();
+
+      // Ensure projects is populated
+      await this.getProjects();
+
+      if (this.supportsFarosClient()) {
```
Comment (chalenge) [inline sources/jira-source/src/project-board-filter.ts:55]: extract this into two functions one handling the `supportsFarosClient`. Hard to follow

File: faros-airbyte-common/src/jira/types.ts [hunk 4]
```diff
@@ -130,3 +130,8 @@ export interface Board extends AgileModels.Board {
   uid: string;
   projectKey: string;
 }
+
+export interface FarosBoard {
+  id: number;
```
Comment (chalenge) [inline faros-airbyte-common/src/jira/types.ts:135]: this is a string if it is coming from faros graph

File: sources/jira-source/src/project-board-filter.ts [hunk 5]
```diff
@@ -48,45 +48,50 @@ export class ProjectBoardFilter {
...
-          await this.maybeIncludeBoard(board.id.toString());
+          // If boards are specified, only include those
+          if (
+            this.config.boards &&
+            !this.config.boards.includes(board.id.toString())
+          ) {
+            continue;
+          }
+          // Check if at least one board project key is in the projects list
+          if (
```
Comment (chalenge) [inline sources/jira-source/src/project-board-filter.ts:69]: if we using `getBoardsFromGraph` only here to get boards then this logic will be much simpler if we invert the query and query for projects and the boards under it
```gql
tms_Project {
  uid
  boards
  ...
}

File: sources/jira-source/src/streams/faros_boards.ts [hunk 6]
```diff
@@ -33,7 +33,7 @@ export class FarosBoards extends StreamWithProjectSlices {
       projectKey: this.projectKey,
     };
 
-    for await (const board of jira.getBoards(this.projectKey)) {
+    for await (const board of await jira.getBoards(this.projectKey)) {
```
Comment (chalenge) [inline sources/jira-source/src/streams/faros_boards.ts:36]: remove extra await, what am I missing?

File: sources/jira-source/src/jira.ts [hunk 7]
```diff
@@ -809,52 +824,58 @@ export class Jira {
...
 
-  async *getBoardsFromGraph(
+  // Fetch project with boards nested from Faros
+  async *getProjectBoardsFromGraph(
     farosClient: FarosClient,
-    graph: string
-  ): AsyncIterableIterator<AgileModels.Board> {
-    const boards = this.iterate(
+    graph: string,
+    projectIds: Array<string>
```
Comment (chalenge) [inline sources/jira-source/src/jira.ts:856]: projectKeys

File: faros-airbyte-common/src/jira/types.ts [hunk 8]
```diff
@@ -139,3 +139,8 @@ export interface IssueProjectVersion {
   readonly key: string;
   readonly projectVersionId: string;
 }
+
+export interface FarosProject {
+  key: string;
+  boardIds: string[];
```
Comment (chalenge) [inline faros-airbyte-common/src/jira/types.ts:145]: boardUids

File: sources/jira-source/src/streams/faros_projects.ts [hunk 9]
```diff
@@ -25,7 +25,7 @@ export class FarosProjects extends StreamBase {
     const projectKeys = this.config.projects?.length
       ? new Set(this.config.projects.map((key) => toUpper(key)))
       : undefined;
-    for await (const project of jira.getProjects(projectKeys)) {
+    for await (const project of await jira.getProjects(projectKeys)) {
```
Comment (chalenge) [inline sources/jira-source/src/streams/faros_projects.ts:28]: remove extra await

File: sources/jira-source/src/jira.ts [hunk 10]
```diff
@@ -809,52 +824,58 @@ export class Jira {
...
           source: 'Jira',
           offset: startAt,
           pageSize: this.maxPageSize,
+          projects: projectIds,
         });
-        return data?.tms_TaskBoard;
+        return data?.tms_Project;
       },
-      async (item: any) => {
+      async (item: any): Promise<FarosProject> => {
```
Comment (chalenge) [inline sources/jira-source/src/jira.ts:868]: this should just be `(item: any): FarosProject`

File: sources/jira-source/src/streams/faros_board_issues.ts [hunk 11]
```diff
@@ -27,7 +27,7 @@ export class FarosBoardIssues extends StreamWithBoardSlices {
     const boardJql = await jira.getBoardJQL(boardConfig.filter.id);
     // Jira Agile API GetConfiguration response type does not include key property
     // but, in practice, it is actually present
-    const projectKey = boardConfig.location['key'];
+    const projectKey = boardConfig.location?.['key'];
```
Comment (chalenge) [inline sources/jira-source/src/streams/faros_board_issues.ts:30]: this then doesn't work for Jira Server right?

Comment (matiaslcoulougian) [reply sources/jira-source/src/streams/faros_board_issues.ts:30]: Yes, projectKey will be null always. I'm wondering, if we are already using board's JQL, why do we need to add the project key condition?

Comment (matiaslcoulougian) [reply sources/jira-source/src/streams/faros_board_issues.ts:30]: I think if the board's JQL does not include a project, board might contain issues from more than one project, so we should not restrict the JQL to only one project (which would be board's location)

**Review by chalenge — APPROVED** [review]
LGTM, after fixing the board JQL
